### ISSUE TITLE
Question about ESP8266 and Arduino

### ISSUE THREAD

**Issue by Sammy1Am** [DJthefirst/Mechanical-Midi-Music#20]
Hey there,

I saw you mention in another issue that this is only working on ESP32s at the moment, and I was curious: is that a library limitation, or is it a performance thing?  Like have you tried running in on ESP8266 or Arduino boards and it didn't work, or is it just not even attemptable yet?

I ask mostly because I _think_ I keep sort of skirting the edges of what ESP8266s can handle, but things tend to 100% work or 100% not-work, so it's hard to tell where the line is and if it's a hardware thing, or I'm just writing iffy code.

**Comment by DJthefirst** [comment]
So it is actually a few different things.

First I haven't fully finished all the configuration code and the GUI so it is a matter of priorities. The current GUI is operational and so is the microcontroller. I just need to add the Distributor configuration to the GUI and finish the built-in HTML midi player, everything works from a DAW. Here is the rough draft of the GUI. A graphic designer friend of mine is mocking up a more professional version so it will have a similar look to Discord. I'll update the moppy project thread sometime next month after my internship finishes in 3 weeks. Also documentation and tutorial videos.
![image](https://github.com/DJthefirst/Mechanical-Midi-Music/assets/37967524/1443a7e1-0fe1-4158-9453-f342d33334a3)

Second I do all my processing on the microcontrollers so my distributors similar to the moppy mapper take up space. I also support up to 32  instruments, with 16 polyphonic notes each per device. I have to go through and optimize my allocation as something like floppy drives are not polyphonic. Currently, the program takes 16.6k bytes of ram and 262k bytes of Flash, Far greater than the Arduino Unos 32k flash and 2k ram. I also want to add name storage and config stuff. How many Arduino IDE libraries I don't need is also something to optimize such as UDP and I2c etc. I plan on making a lite version for most use cases but I want to finish building up before I strip back. I have gotten Moppy running on an Arduino Nano before so that is my target.

My current Code is also C++ 17, not 11 so I have to rewrite some of the optimized functions.

For performance, I haven't done too much testing. I have tests planned to see how low I can get the Timer resolution before the ISR overruns and everything crashes. I also need to finish building out my platform layer where I can optimize the config for each microcontroller. Also, need to add more support for things such as the STM 32 Interrupt and other platforms.

I just got two of the new Esp32 boards and want to see if I can port a version of this code to the Espressif IDF. Then I can achieve USB host support and every device will act as a USB Midi Device instead of a Serial Port. I have to decide if I want to prioritize that over expanding to other platforms.

TLDR. My summer internship ends in 3 weeks, I am prioritizing finishing the GUI, I need to optimize memory usage, and need to finish porting other platforms.

**Comment by DJthefirst** [comment]
Just heard back about the new GUI still need to code it up but this is the reference.
![MIDI Screen](https://github.com/DJthefirst/Mechanical-Midi-Music/assets/37967524/826cdaa3-ea9a-4b2f-b368-e220b777b70a)

**Comment by Sammy1Am** [comment]
(Finally cleaning out my inbox after way too long letting it languish, apologies)

Thanks for the details, and the GUI looks awesome.  I really think what you're aiming for will be a much more flexible/portable solution than Moppy.

I can see how moving a lot of the configuration and mapping to the microcontroller would start to eat up memory, so the ESP32 limitation makes sense.  I've probably mentioned before, but my very first attempt I tried to convert MIDI messages on the Arduino itself, but it could only handle one note at a time before the effort of decoding the MIDI would break the loop driving the steppers, **however**, I didn't know about interrupts at the time, so everything was just crammed in `loop()`.  It's left me wondering if actually everything could have been on the Arduino if I'd know about interrupts.  But on the other hand, like you're seeing, all the mapping and stuff that can easily be done on the PC might have taken a lot more optimization to fit on the MC.

Keep up the good work; someday if I ever have time to dust off the drives I'll have to give your code a spin :)
